src/raftstore/coprocessor/region_info_accessor.rs

```diff
@@ -0,0 +1,753 @@
+        };
+        self.scheduler
+            .schedule(RegionCollectionMsg::RaftStoreEvent(event))
+            .unwrap();
+    }
+}
+
+impl RoleObserver for RegionEventListener {
+    fn on_role_change(&self, context: &mut ObserverContext, role: StateRole) {
+        let region = context.region().clone();
+        let event = RaftStoreEvent::RoleChange { region, role };
+        self.scheduler
+            .schedule(RegionCollectionMsg::RaftStoreEvent(event))
+            .unwrap();
+    }
+}
+
+/// Create an `RegionEventListener` and register it to given coprocessor host.
+fn register_region_event_listener(
+    host: &mut CoprocessorHost,
+    scheduler: Scheduler<RegionCollectionMsg>,
+) {
+    let listener = RegionEventListener { scheduler };
+
+    host.registry
+        .register_role_observer(1, box listener.clone());
+    host.registry
+        .register_region_change_observer(1, box listener);
+}
+
+/// `RegionCollection` is the place where we hold all region information we collected, and the
+/// underlying runner of `RegionInfoAccessor`. It listens on events sent by the `RegionEventListener` and
+/// keeps information of all regions. Role of each region are also tracked.
+struct RegionCollection {
+    // HashMap: region_id -> (Region, State)
+    regions: RegionsMap,
+    // BTreeMap: 'z' + end_key -> region_id
+    region_ranges: RegionRangesMap,
+}
+
+impl RegionCollection {
+    fn new() -> Self {
+        Self {
+            regions: HashMap::default(),
+            region_ranges: BTreeMap::default(),
+        }
+    }
+
+    /// Check if the end_key has collided with another region. If so, try to clean it.
+    /// If the collided region has a greater version than the current one, it means that the current
+    /// operation is caused by a stale message. Then this function should fail.
+    ///
+    /// Returns a bool value indicating whether checking succeeded.
+    fn check_exist(&mut self, region_id: u64, version: u64, end_key: &[u8]) -> bool {
+        if let Some(collided_region_id) = self.region_ranges.get(end_key).cloned() {
+            // There is already another region with the same end_key
+            assert_ne!(collided_region_id, region_id);
+
+            let collided_region_version = self.regions[&collided_region_id]
+                .region
+                .get_region_epoch()
+                .get_version();
+            if version < collided_region_version {
+                // Another region is the actual new region. Now we are trying to create a region
+                // because of an old message. Fail.
+                return false;
+            }
+            // Another region is older. Remove it.
+            info!(
+                "region_collection: remove region {} because colliding with region {}",
+                collided_region_id, region_id
+            );
+            self.regions.remove(&collided_region_id);
```

Comment:
Also need to remove the related information in region_ranges.

Reply:
We will insert another region at that key. It will be replaced.

Reply:
Oh, I know what you means.